README.md
# ORIZIN UI

日本語 | [English](README_en.md)

![Broken Link Check](https://github.com/Robot-Inventor/ORIZIN-UI/workflows/Broken%20Link%20Check/badge.svg) ![CodeQL](https://github.com/Robot-Inventor/ORIZIN-UI/workflows/CodeQL/badge.svg) [![deepcode](https://www.deepcode.ai/api/gh/badge?key=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwbGF0Zm9ybTEiOiJnaCIsIm93bmVyMSI6IlJvYm90LUludmVudG9yIiwicmVwbzEiOiJPUklaSU4tVUkiLCJpbmNsdWRlTGludCI6ZmFsc2UsImF1dGhvcklkIjoxOTUwNCwiaWF0IjoxNjAxNDgwMDM3fQ.6SSplvJLANbclZT-dLtCYyz-tYTmFlI5SFwiWuC1Ykc)](https://www.deepcode.ai/app/gh/Robot-Inventor/ORIZIN-UI/_/dashboard?utm_content=gh%2FRobot-Inventor%2FORIZIN-UI)

ORIZINシリーズの開発過程でできたUI関連のライブラリーです。Google Chrome 86で動作確認しています。古いGoogle Chromeのバージョンや他のWEBブラウザーでは正常に動作しない可能性があります。Google Chrome以外のWEBブラウザーにも対応する必要がある場合は必ず事前に動作確認を行ってください。

## できること

ORIZIN UIを使用すると、以下のものを利用できるようになります。

- トグルスイッチ
- テキスト入力中に色が変わるアンダーライン付きテキストボックス
- マテリアルデザイン風リップルエフェクト
- 画面内の通知の表示

## デモ

[https://robot-inventor.github.io/ORIZIN-UI/index.html](https://robot-inventor.github.io/ORIZIN-UI/index.html)

## 使い方

ORIZIN UIを使った簡単な例です。

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>ORIZIN UI Sample</title>
</head>
<body>
    <toggle-switch class="ripple_effect"></toggle-switch>    <!-- トグルスイッチ -->
    <underlined-textbox></underlined-textbox>    <!-- アンダーライン付きテキストボックス -->
</body>
<script src="orizin_ui.min.js"></script>
<script>
    new Ripple(".ripple_effect", {
        debug: false, // Ripple.jsのロギングのon/of
        on: 'mousedown', // リップルエフェクトのイベントトリガー

        opacity: 0.4, // リップルの不透明度
        color: "auto", // 背景色の設定。"auto"の場合はテキストカラーを使用
        multi: false, // 同じ要素内に同時に複数のリップルを生成することを許可

        duration: 0.7, // リップルの持続時間

        // リップルの速度を変更するためのフィルタ機能
        rate: function(pxPerSecond) {
            return pxPerSecond;
        },

        easing: 'linear'
    });

    const notifier = new Notification();
    notifier.notify({
        message: "Notification test", // 通知の内容
        callback: function() {  // 通知が閉じられたときに実行する処理
            alert("Notification closed.");
        },
        auto_close_time: 10000, // 自動で通知を閉じるまでの時間（ミリ秒）
        type = "normal", // 通知の種類。「normal」「caution」「warning」から選択可能
    });
</script>
</html>
```

### リップルエフェクトについて

指定した要素に対してマテリアルデザイン風のリップルエフェクトを提供します。

内部的には、リップルエフェクトを生成するjQueryプラグインの[jakiestfu/Ripple.js](https://github.com/jakiestfu/Ripple.js)をjQuery非依存かつ1個のJavaScriptファイルのみで動作するように書き換えたものを使用しています。

### 画面内通知について

画面の右下に通知を表示する機能を提供します。たとえば以下のようにして使います。

```javascript
const notifier = new Notification();
notifier.notify({
    message: "Notification test", // 通知の内容
    callback: function() {  // 通知が閉じられたときに実行する処理
        alert("Notification closed.");
    },
    auto_close_time: 10000, // 自動で通知を閉じるまでの時間（ミリ秒）
    type = "normal", // 通知の種類。「normal」「caution」「warning」から選択可能
});
```

すべての引数は省略可能で、デフォルト値は以下のとおりです。

|引数|デフォルト値|
|:--|:--|
|message|（空）|
|callback|（なし）|
|auto_close_time|undefined（自動で閉じない）|
|type|normal|

--------------------

ここから先はトグルスイッチとアンダーライン付きテキストボックスの説明になります。

### CSSの適用方法

``::part()``を使用してスタイルを適用できます。

#### トグルスイッチの場合

トグルスイッチの場合は以下のようにしてスタイルを適用してください。

```css
toggle-switch::part(foundation) {
    /* 通常時のトグルスイッチの土台部分のスタイル */
}

toggle-switch[checked]::part(foundation) {
    /* onのときのトグルスイッチの土台部分のスタイル */
}

toggle-switch::part(handle) {
    /* 通常時のトグルスイッチのハンドル部分のスタイル */
}

toggle-switch[checked]::part(handle) {
    /* onのときのトグルスイッチのハンドル部分のスタイル */
}

toggle-switch::part(ripple) {
    /* トグルスイッチのリップルエフェクトのスタイル */
}

toggle-switch:active::part(ripple) {
    /* トグルスイッチのクリック時のリップルエフェクトのスタイル */
}
```

注意点として、``toggle-switch``そのものに``width``や``height``を指定してもトグルスイッチの大きさは変わりません。大きさを変更したい場合は基本的には上記のCSSの書き方でパーツごとに行うことになりますが、縦と横の比率を変えないのであれば``transform``で十分かもしれません。たとえば2倍の大きさにする場合は以下のようになります。

```css
toggle-switch {
    transform: scale(2);
}
```

#### アンダーライン付きテキストボックスの場合

アンダーライン付きテキストボックスの場合は以下のようにしてスタイルを適用してください。

```css
underlined-textbox::part(textbox) {
    /* 通常時のテキスト入力部分のスタイル */
}

underlined-textbox::part(textbox):focus {
    /* テキスト入力中のテキスト入力部分のスタイル */
}

underlined-textbox::part(normal_underline) {
    /* 通常時のアンダーラインのスタイル */
}

underlined-textbox::part(focused_underline) {
    /* テキスト入力中のアンダーラインのスタイル */
}
```

### 属性

トグルスイッチは内部的には通常のinput要素のチェックボックス、アンダーライン付きテキストボックスは内部的には通常のinput要素のテキストボックスを利用しています。トグルスイッチやアンダーライン付きテキストボックスに与えられた属性は内部で使用しているinput要素にそのまま渡されます。トグルスイッチとアンダーライン付きテキストボックスが対応している属性は以下のとおりです。

#### トグルスイッチが対応している属性

- checked

#### アンダーライン付きテキストボックスが対応している属性

- value
- autocomplete
- list
- maxlength
- minlength
- pattern
- placeholder
- readonly
- required
- size

これらの属性はJavaScriptの``getAttribute()``や``setAttribute()``で取得/設定できます。また、JavaScriptプロパティーを使用して取得/設定することもできます。

#### イベント

changeイベントとinputイベントに対応しています。イベントの発火条件はトグルスイッチはinput要素のチェックボックス、アンダーライン付きテキストボックスはinput要素のテキストボックスと同じです。これは、内部で使用しているinput要素のイベントをそのまま伝えているためです。

## ORIZINシリーズとは

ORIZINシリーズは、Robot-Inventorが開発したさまざまなソフトウェアです。

↓一覧

|名称|概要|
|:--|:--|
|ORIZIN Agent HTML|Pythonで作られたオープンソースのAIアシスタント。キャッチコピーは「いつもあなたと共に。オープンな相棒」|
|ORIZIN Paon|ホラーゲームの「ぴえん」に触発されたWebGLを使用したオープンソースのホラーゲーム|

## OSSライセンス

本ソフトウェアでは、以下のOSSを使用しています。OSSの名前をクリックするとライセンスが表示されます。

<details>
    <summary>ORIZIN Agent HTML</summary>
MIT License

Copyright (c) 2019 - 2020 Robot-Inventor

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
</details>
<details>
    <summary>Ripple.js v1.2.1</summary>
The MIT License (MIT)

Copyright (c) 2014 Jacob Kelley

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
</details>


changelog.md
# Change Log

このプロジェクトに対するすべての重要な変更は、このファイルに文書化されます。

## 情報

Series Name: ORIZIN UI
Edition: N/A

他の情報を読むには、information.txtを参照してください。

## バージョニング

"a.b.c.de-f"というバージョン番号の意味は以下の通りです。

|位置|意味|文字の種類|最小値|
|:--:|:--:|:--:|:--:|
|a|メジャーバージョン|数字|0|
|b|マイナーバージョン|数字|0|
|c|ビルド（累積バージョン）|数字|1|
|d|リビジョン|数字|0|
|e|リリース|アルファベット（devまたはなし）|N/A|
|f|開発コードネーム|アルファベット|N/A|

それぞれの言葉の意味は以下の通りです。

|言葉|意味|
|:--:|:--:|
|メジャーバージョン|大きな機能の追加またはUIの大きな変更|
|マイナーバージョン|小さな機能の追加またはUIの小さな変更|
|ビルド（累積バージョン）|バージョン番号が変わるたびにこの数字が増えます|
|リビジョン|バグの修正またはプログラム中のコメントの変更|
|リリース|プログラムがDevelop版の場合"dev"になる|
|開発コードネーム|メジャー（マイナー）バージョンのコードネーム|

## リリース

### [v4.1.11.1-Duck] - 2020-10-26

#### 変更

- トグルスイッチのリップルエフェクトの不透明度を上げて見やすくしました。
- サンプルページにおいてAndroidのGoogle Chromeのコンテンツのタップ時に表示される水色の四角を非表示に変更しました。

### [v4.1.10.0-Duck] - 2020-10-26

#### 追加

- トグルスイッチにリップルエフェクトを追加しました。

#### 変更

- AndroidのGoogle Chromeでトグルスイッチをクリックしたときに表示される水色の四角を非表示にしました。

### [v4.0.9.2-Duck] - 2020-10-25

#### 修正

- サンプルページのリップルエフェクトの調整を行いました。

### [v4.0.8.1-Duck] - 2020-10-25

#### 変更

- サンプルページのデザインを改善しました。
- サンプルページの言語切替をページ変遷なしで行えるようになりました。

### [v4.0.7.0-Duck] - 2020-10-24

#### 追加

- Notification機能を追加しました。

### [v3.0.6.1-Crane] - 2020-10-15

#### 追加

- changelog.mdを追加しました。

#### 修正

- underlined-textboxのvalueを変更しても表示に反映されないバグを修正しました。

### [v3.0.5.0-Crane] - 2020-10-05

#### 追加

- [jakiestfu/Ripple.js](https://github.com/jakiestfu/Ripple.js)をjQuery非依存かつ1個のJavaScriptファイルのみで動作するように書き換えたものを統合し、リップルエフェクト機能を使用できるようになりました。

### [v2.0.4.0-Budgerigar] - 2020-10-04

#### 変更

- Web Componentsを使用したものに変更し、より使いやすくしました。以前のバージョンとの互換性はありません。

### [v1.1.3.1-Albatross] - 2020-10-02

#### 追加

- orizin_ui.jsをMinify化したorizin_ui.min.jsを追加しました。

### [v1.1.2.0-Albatross] - 2020-10-01

#### 変更

- カスタマイズの方法をCSS変数を用いる方法から直接関数に渡す方法に変更

### [v1.0.1.0-Albatross] - 2020-10-01

#### 追加

- 最初のバージョン


sample/main.js
function parse_url_parameter() {
    const url_parameters_list = location.search.substring(1).split("&");
    const url_parameters = {};
    for(let i = 0; i < url_parameters_list.length; i++) {
        const parsed_parameter = url_parameters_list[i].split("=");
        url_parameters[parsed_parameter[0]] = parsed_parameter[1];
    }
    return url_parameters;
}

function normalize_language_setting(language) {
    const browser_language = navigator.language;
    const default_language = browser_language ==="ja" ? "ja" : language === "en" ? "en" : "ja";
    return language === "ja" ? "ja" : language === "en" ? "en" : default_language;
}

function load_texts(language) {
    const language_setting = normalize_language_setting(language);

    function return_text(object) {
        return language_setting === "ja" ? object["ja"] : language_setting === "en" ? object["en"] : object["ja"];
    }

    const h1 = document.querySelector("h1");
    const h1_text = {
        ja: "ORIZIN UIサンプル",
        en: "ORIZIN UI Sample"
    }
    h1.textContent = return_text(h1_text);

    const download_link = document.getElementById("download_link");
    const download_link_text = {
        ja: "ダウンロード",
        en: "Download"
    }
    download_link.textContent = return_text(download_link_text);

    const h2_list = document.querySelectorAll("h2");
    const h2_texts = {
        ja: ["トグルスイッチ", "アンダーライン付きテキストボックス", "リップルエフェクト", "通知"],
        en: ["Toggle Switch", "Underlined Textbox", "Ripple Effect", "Notification"]
    };
    for(let i = 0; i < h2_list.length; i++) {
        h2_list[i].textContent = return_text(h2_texts)[i];
    }

    const button_list = document.querySelectorAll("button");
    const button_texts = {
        ja: ["値を取得", "通知を表示"],
        en: ["Get value", "Notify"]
    };
    for(let i = 0; i < button_list.length; i++) {
        button_list[i].textContent  = return_text(button_texts)[i];
    }

    const multi_language_support_element = document.querySelectorAll(".multi_language_support_text");
    const multi_language_support_text = {
        ja: ["結果：", "クリックしてください", "ダブルクリックしてください", "通常", "注意", "警告"],
        en: ["Result: ", "Click me!", "Double click me!", "normal", "caution", "warning"]
    }
    for(let i = 0; i < multi_language_support_element.length; i++) {
        multi_language_support_element[i].textContent = return_text(multi_language_support_text)[i];
    }
}

let language = parse_url_parameter().lang || "ja";
document.getElementById("language_select").language.value = normalize_language_setting(language);
load_texts(language);

document.getElementById("language_select").addEventListener("change", function() {
    history.replaceState(null, null, "?lang=" + document.getElementById("language_select").language.value);
    load_texts(parse_url_parameter().lang);
});


const checkbox = document.getElementById("checkbox_sample");
checkbox.addEventListener("change", function() {
    let checkbox_status;
    if(checkbox.checked) {
        checkbox_status = "checked!";
    } else {
        checkbox_status = "not checked!";
    }
    document.getElementById("checkbox_result").textContent = checkbox_status;
});

const checkbox2 = document.getElementById("checkbox_sample2");
checkbox2.addEventListener("change", function() {
    let checkbox_status;
    if(checkbox2.checked) {
        checkbox_status = "checked!";
    } else {
        checkbox_status = "not checked!";
    }
    document.getElementById("checkbox_result2").textContent = checkbox_status;
});

document.getElementById("get_text").addEventListener("click", function() {
    document.getElementById("underline_textbox_result").textContent = document.getElementById("underline_textbox_sample").value;
});

new Ripple(".ripple_effect");

new Ripple(".multi_ripple_effect", {
    multi: true
});

const notifier = new Notification();

document.getElementById("notify_button").addEventListener("click", function() {
    notifier.notify({
        message: document.getElementById("notify_content_input").value,
        type: document.getElementById("notification_setting").notification_type.value
    });
});


orizin_ui.js
class ToggleSwitch extends HTMLElement {
    static get observedAttributes() {
        return ["checked"];
    }

    constructor() {
        super();

        this._checked = null;

        const toggle_width = "2.5rem";
        const toggle_height = "1rem";
        const toggle_border_thickness = "0.05rem";
        const toggle_border_color = "gray";
        const toggle_disabled_color = "white";
        const toggle_enabled_color = "#adff99";
        const toggle_handle_color = "white";
        const toggle_handle_diameter = `calc(${toggle_height} + 0.4rem)`;

        const shadow = this.attachShadow({mode: "closed"});

        const checkbox_element = document.createElement("input");
        this._checkbox_element = checkbox_element;
        checkbox_element.setAttribute("type", "checkbox");
        checkbox_element.setAttribute("id", "checkbox");
        if(this.hasAttribute("checked")) {
            checkbox_element.checked = true;
        }

        const toggle_background_element = document.createElement("label");
        toggle_background_element.setAttribute("for", "checkbox");
        toggle_background_element.setAttribute("part", "foundation");

        const toggle_handle_element = document.createElement("div");
        toggle_handle_element.setAttribute("part", "handle");

        const ripple = document.createElement("div");
        ripple.setAttribute("part", "ripple");
        ripple.setAttribute("part", "ripple");
        toggle_handle_element.appendChild(ripple);

        const style = document.createElement("style");
        style.textContent = `
#checkbox {
    display: none;
    transition: 0.3s;
}

#checkbox+label {
    display: inline-block;
    background: ${toggle_disabled_color};
    width: ${toggle_width};
    height: ${toggle_height};
    border: solid ${toggle_border_thickness} ${toggle_border_color};
    border-radius: ${toggle_width};
    margin: calc(${toggle_handle_diameter} - ${toggle_height});
    padding: 0;
    vertical-align: bottom;
    overflow: visible;
    transition: 0.3s;
    cursor: pointer;
    -webkit-tap-highlight-color: transparent;
}

#checkbox:checked+label {
    background: ${toggle_enabled_color};
}

#checkbox+label div {
    display: block;
    background: ${toggle_handle_color};
    width: ${toggle_handle_diameter};
    height: ${toggle_handle_diameter};
    border: ${toggle_border_color} solid ${toggle_border_thickness};
    border-radius: ${toggle_width};
    transition: 0.3s;
    margin: 0;
    position: relative;
    top: calc(${toggle_border_thickness} * -1 - 0.2rem);
    left: calc(${toggle_border_thickness} * -1 - 0.2rem);
}

#checkbox:checked+label div {
    margin-left: 100%;
    transform: translateX(-50%);
}

#checkbox+label div div {
    content: "";
    position: absolute;
    width: 200%;
    height: 200%;
    top: -50%;
    left: -50%;
    background: gray;
    border-radius: 100%;
    opacity: 0;
    transition: 0.3s;
}

#checkbox+label div:active div,
#checkbox+label:active div div {
    opacity: 0.7;
}

:host {
    display: inline-block;
}
        `;

        shadow.appendChild(checkbox_element);
        shadow.appendChild(toggle_background_element);
        shadow.appendChild(style);
        toggle_background_element.appendChild(toggle_handle_element);

        checkbox_element.addEventListener("change", () => {
            if(checkbox_element.checked) {
                this.setAttribute("checked", "");
            } else {
                this.removeAttribute("checked");
            }
            const change_event = new CustomEvent("change");
            this.dispatchEvent(change_event);
        });

        checkbox_element.addEventListener("input", () => {
            if(checkbox_element.checked) {
                this.setAttribute("checked", "");
            } else {
                this.removeAttribute("checked");
            }
        });
    }

    attributeChangedCallback(name, old_value, new_value) {
        switch(name) {
            case "checked":
                this._checked = new_value;
                if(new_value || new_value === "") {
                    this._checkbox_element.setAttribute("checked", "");
                } else {
                    this._checkbox_element.removeAttribute("checked");
                }
                break;
        }
    }

    get checked() {
        return this.hasAttribute("checked");
    }

    set checked(value) {
        if(value) {
            this._checkbox_element.setAttribute("checked", "");
            this.setAttribute("checked", "");
        } else {
            this._checkbox_element.removeAttribute("checked");
            this.removeAttribute("checked");
        }
    }
}

customElements.define("toggle-switch", ToggleSwitch);


class UnderlineTextbox extends HTMLElement {
    static get observedAttributes() {
        return ["value", "autocomplete", "list", "maxlength", "minlength", "pattern", "placeholder", "readonly", "required", "size"];
    }

    constructor() {
        super();

        this._value = null;
        this._autocomplete = null;
        this._list = null;
        this._maxLength = null;
        this._minLength = null;
        this._pattern = null;
        this._placeholder = null;
        this._readOnly = null;
        this._required = null;
        this._size = null;

        const underline_color_normal = "rgba(0, 0, 0, 0.7)";
        const underline_color_focused = "#ff6a00";
        const underline_thickness = "0.15rem";

        const shadow = this.attachShadow({mode: "closed"});

        const input_element = document.createElement("input");
        this._input_element = input_element;
        input_element.setAttribute("type", "text");
        input_element.setAttribute("id", "textbox");
        input_element.setAttribute("part", "textbox");

        if(this.hasAttribute("value")) {
            input_element.value = this.getAttribute("value");
        }
        if(this.hasAttribute("autocomplete")) {
            input_element.setAttribute("autocomplete", this.getAttribute("autocomplete"));
        }
        if(this.hasAttribute("list")) {
            input_element.setAttribute("list", this.getAttribute("list"));
        }
        if(this.hasAttribute("maxlength")) {
            input_element.setAttribute("maxlength", this.getAttribute("maxlength"));
        }
        if(this.hasAttribute("minlength")) {
            input_element.setAttribute("minlength", this.getAttribute("minlength"));
        }
        if(this.hasAttribute("pattern")) {
            input_element.setAttribute("pattern", this.getAttribute("pattern"));
        }
        if(this.hasAttribute("placeholder")) {
            input_element.setAttribute("placeholder", this.getAttribute("placeholder"));
        }
        if(this.hasAttribute("readonly")) {
            input_element.setAttribute("readonly", this.getAttribute("readonly"));
        }
        if(this.hasAttribute("required")) {
            input_element.setAttribute("required", this.getAttribute("required"));
        }
        if(this.hasAttribute("size")) {
            input_element.setAttribute("size", this.getAttribute("size"));
        }

        const underline_group = document.createElement("div");
        underline_group.setAttribute("id", "underline_group");

        const normal_underline = document.createElement("div");
        normal_underline.setAttribute("id", "normal_underline");
        normal_underline.setAttribute("part", "normal_underline");

        const focused_underline = document.createElement("div");
        focused_underline.setAttribute("id", "focused_underline");
        focused_underline.setAttribute("part", "focused_underline");

        const style = document.createElement("style");
        style.textContent = `
#textbox {
    width: 100%;
    height: 100%;
    border: none;
    outline: none;
    background: transparent;
    margin: 0 0 ${underline_thickness} 0;
    padding: ${underline_thickness};
    box-sizing: border-box;
}

#textbox + #underline_group {
    width: 100%;
    height: ${underline_thickness};
    position: relative;
    left: 0;
    top: calc(-1.5 * ${underline_thickness});
    margin: 0;
    padding: 0;
}

#textbox + #underline_group #normal_underline {
    width: 100%;
    height: 100%;
    position: absolute;
    background: ${underline_color_normal};
    border-radius: ${underline_thickness};
    margin: 0;
    padding: 0;
}

#textbox + #underline_group #focused_underline {
    width: 0;
    height: 100%;
    position: absolute;
    background: ${underline_color_focused};
    transition: 0.3s;
    margin: 0 0 0 50%;
    padding: 0;
    border-radius: ${underline_thickness};
}

#textbox:focus + #underline_group #focused_underline {
    width: 100%;
    margin: 0;
}

:host {
    display: inline-block;
}
        `;

        shadow.appendChild(input_element);
        underline_group.appendChild(normal_underline);
        underline_group.appendChild(focused_underline);
        shadow.appendChild(underline_group);
        shadow.appendChild(style);

        input_element.addEventListener("change", () => {
            this.setAttribute("value", input_element.value);
            const change_event = new CustomEvent("change");
            this.dispatchEvent(change_event);
        });

        input_element.addEventListener("input", () => {
            this.setAttribute("value", input_element.value);
        });
    }

    get value() {
        return this._value;
    }
    set value(v) {
        this.setAttribute("value", v);
    }

    get autocomplete() {
        return this._autocomplete;
    }
    set autocomplete(v) {
        this.setAttribute("autocomplete", v);
    }

    get list() {
        return this._list;
    }
    set list(v) {
        this.setAttribute("list", v);
    }

    get maxLength() {
        return this._maxLength;
    }
    set maxLength(v) {
        this.setAttribute("maxlength", v);
    }

    get minLength() {
        return this._minLength;
    }
    set minLength(v) {
        this.setAttribute("minlength", v);
    }

    get pattern() {
        return this._pattern;
    }
    set pattern(v) {
        this.setAttribute("pattern", v);
    }

    get placeholder() {
        return this._placeholder;
    }
    set placeholder(v) {
        this.setAttribute("placeholder", v);
    }

    get readOnly() {
        return this._readOnly;
    }
    set readOnly(v) {
        this.setAttribute("readOnly", v);
    }

    get required() {
        return this._required;
    }
    set required(v) {
        this.setAttribute("required", v);
    }

    get size() {
        return this._size;
    }
    set size(v) {
        this.setAttribute("size", v);
    }

    attributeChangedCallback(name, old_value, new_value) {
        switch(name) {
            case "value":
                this._value = this._input_element.value = new_value;
                break;
            case "autocomplete":
                this._autocomplete = this._input_element.autocomplete = new_value;
                break;
            case "list":
                this._list = this._input_element.list = new_value;
                break;
            case "maxlength":
                this._maxLength = this._input_element.maxLength = new_value;
                break;
            case "minlength":
                this._minLength = this._input_element.minLength = new_value;
                break;
            case "pattern":
                this._pattern = this._input_element.pattern = new_value;
                break;
            case "placeholder":
                this._placeholder = this._input_element.placeholder = new_value;
                break;
            case "readonly":
                this._readOnly = this._input_element.readOnly = new_value;
                break;
            case "required":
                this._required = this._input_element.required = new_value;
                break;
            case "size":
                this._size = this._input_element.size = new_value;
                break;
        }
    }
}

customElements.define("underlined-textbox", UnderlineTextbox);

class Ripple {
    /*! The Original Program: 
    *
    * Ripple.js v1.2.1
    * The MIT License (MIT)
    * Copyright (c) 2014 Jacob Kelley
    *
    * Improved by Copyright (c) 2020 Robot-Inventor */
    constructor(selector, options) {
        const self = this;
        const ripple_style = document.createElement("style");
        ripple_style.textContent = `
.has-ripple{
    position: relative;
    overflow: hidden;
    -webkit-transform: translate3d(0,0,0);
    -o-transform: translate3d(0,0,0);
    transform: translate3d(0,0,0);
}

.ripple{
    display: block;
    position: absolute;
    pointer-events: none;
    border-radius: 50%;
    -webkit-transform: scale(0);
    -o-transform: scale(0);
    transform: scale(0);
    background: #fff;
    opacity: 1;
}

.ripple-animate{
    -webkit-animation: ripple;
    -o-animation: ripple;
    animation: ripple;
}

@-webkit-keyframes ripple{
    100%{
        opacity: 0;
        -webkit-transform: scale(2);
        transform: scale(2);
    }
}

@-o-keyframes ripple{
    100%{
        opacity: 0;
        -o-transform: scale(2);
        transform: scale(2);
    }
}

@keyframes ripple{
    100%{
        opacity: 0;
        transform: scale(2);
    }
}
        `;
        document.body.appendChild(ripple_style);

        const _log = self.log = function() {
            if(self.defaults.debug && console && console.log) {
                console.log.apply(console, arguments);
            }
        };

        self.selector = selector;
        self.defaults = {
            debug: false,
            on: "mousedown",

            opacity: 0.4,
            color: "auto",
            multi: false,

            duration: 0.7,
            rate: function(pxPerSecond) {
                return pxPerSecond;
            },

            easing: "linear"
        };

        self.defaults = Object.assign({}, self.defaults, options);

        function Trigger(e) {

            let ripple;

            this.classList.add("has-ripple");

            // This instances settings
            const settings = Object.assign({}, self.defaults, this.dataset);

            // Create the ripple element
            if (settings.multi || (!settings.multi && this.querySelectorAll(".ripple").length === 0)) {
                ripple = document.createElement("span");
                ripple.classList.add("ripple");
                this.appendChild(ripple);

                _log("Create: Ripple");

                // Set ripple size
                if (!ripple.clientHeight && !ripple.clientWidth) {
                    const size = Math.max(this.offsetWidth, this.offsetHeight);
                    ripple.style.height = size + "px";
                    ripple.style.width = size + "px";
                    _log("Set: Ripple size");
                }

                // Give the user the ability to change the rate of the animation
                // based on element width
                if(settings.rate && typeof settings.rate == "function") {

                    // rate = pixels per second
                    const rate = Math.round(ripple.clientWidth / settings.duration);

                    // new amount of pixels per second
                    const filteredRate = settings.rate(rate);

                    // Determine the new duration for the animation
                    const newDuration = (ripple.clientWidth / filteredRate);

                    // Set the new duration if it has not changed
                    if(settings.duration.toFixed(2) !== newDuration.toFixed(2)) {
                        _log("Update: Ripple Duration", {
                            from: settings.duration,
                            to: newDuration
                        });
                        settings.duration = newDuration;
                    }
                }

                // Set the color and opacity
                const color = (settings.color == "auto") ? getComputedStyle(this, null).getPropertyValue("color") : settings.color;
                const css = {
                    animationDuration: (settings.duration).toString() + "s",
                    animationTimingFunction: settings.easing,
                    background: color,
                    opacity: settings.opacity
                };

                _log("Set: Ripple CSS", css);
                ripple.style.animationDuration = (settings.duration).toString() + "s";
                ripple.style.animationTimingFunction = settings.easing;
                ripple.style.background = color;
                ripple.style.opacity = settings.opacity;
            }

            // Ensure we always have the ripple element
            if(!settings.multi) {
                _log("Set: Ripple Element");
                ripple = this.querySelector(".ripple");
            }

            // Kill animation
            _log("Destroy: Ripple Animation");
            ripple.classList.remove("ripple-animate");


            // Retrieve coordinates
            const x = e.pageX - this.getBoundingClientRect().left - document.documentElement.scrollLeft - ripple.clientWidth / 2;
            const y = e.pageY - this.getBoundingClientRect().top - document.documentElement.scrollTop - ripple.clientHeight / 2;

            /**
             * We want to delete the ripple elements if we allow multiple so we dont sacrifice any page
             * performance. We don't do this on single ripples because once it has rendered, we only
             * need to trigger paints thereafter.
             */
            if(settings.multi) {
                _log("Set: Ripple animationend event");
                ["animationend", "webkitAnimationEnd", "oanimationend", "MSAnimationEnd"].forEach(function(e) {
                    ripple.addEventListener(e, function() {
                        _log("Note: Ripple animation ended");
                        _log("Destroy: Ripple");
                        ripple.remove();
                    }, {
                        once: true
                    });
                });
            }

            // Set position and animate
            _log("Set: Ripple location");
            _log("Set: Ripple animation");
            ripple.style.top = y + "px";
            ripple.style.left = x + "px";
            ripple.classList.add("ripple-animate");
        }

        const ripple_elements = document.querySelectorAll(self.selector);
        ripple_elements.forEach(element => {
            element.addEventListener(self.defaults.on, Trigger);
        });
    }
}

class Notification {
    constructor() {
        const notification_style = document.createElement("style");
        notification_style.textContent = `
div#notification_outer {
    width: min-content;
    position: fixed;
    bottom: 0.5rem;
    right: 0.5rem;
    text-align: right;
}

div.notification {
    width: 50vmin;
    height: auto;
    display: inline-block;
    padding: 0.75rem;
    margin-top: 0.5rem;
    transition: 0.3s;
    box-shadow: 0.2rem 0.2rem 1rem rgba(0, 0, 0, 0.3);
    animation: notification_animation 1s ease;
    cursor: default;
    text-align: left;
}

div.notification:hover {
    box-shadow: 0.2rem 0.2rem 2rem rgba(0, 0, 0, 0.3);
}

div.notification.normal {
    background: rgb(255, 255, 255);
    border-bottom: solid 0.1rem blue;
    color: #959595;
}

div.notification.warning {
    background: rgb(255, 220, 220);
    border-bottom: solid 0.1rem red;
    color: #838382;
}

div.notification.caution {
    background: rgb(255, 255, 220);
    border-bottom: solid 0.1rem rgb(255, 180, 0);
    color: #949393;
}

div.notification span.close {
    cursor: pointer;
    font-weight: bold;
    margin-right: 1rem;
    vertical-align: top;
    user-select: none;
}

div.notification p {
    width: auto;
    max-width: calc(100% - 3rem);
    height: auto;
    display: inline-block;
    margin: 0;
    padding: 0;
}

@keyframes notification_animation {
    0% {
        opacity: 0;
        transform: translate(100%);
    }
    100% {
        opacity: 1;
        transform: none;
    }
}
        `;
        document.body.appendChild(notification_style);

        const notification_outer = document.createElement("div");
        notification_outer.setAttribute("id", "notification_outer");
        document.body.appendChild(notification_outer);
    }

    notify({
        message = "",
        callback = function(){},
        auto_close_time,
        type = "normal",
    } = {}
    ) {
        function close_notify() {
            notifier.style.opacity = 0;
            callback();
            setTimeout(() => {
                notifier.remove();
            }, 300);
        }

        const notifier = document.createElement("div");
        notifier.setAttribute("class", `notification ${type}`);

        const close_button = document.createElement("span");
        close_button.setAttribute("class", "close");
        close_button.setAttribute("title", "close");
        close_button.textContent = "X";
        close_button.addEventListener("click", function() {
            close_notify();
        });

        const message_area = document.createElement("p");
        message_area.textContent = message;

        notifier.appendChild(close_button);
        notifier.appendChild(message_area);

        document.getElementById("notification_outer").appendChild(notifier);
        setTimeout(function() {
            if(auto_close_time) {
                setTimeout(function() {
                    close_button.click();
                }, auto_close_time);
            }
        }, 300);
    }
}
